nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -4,97 +4,218 @@
+                    warn!(log, "{err:#}");
+                    return Some(err);
+                }
+            };
+
+            let client = nexus_networking::sled_client_from_address(
+                sled_id.into_untyped_uuid(),
+                db_sled.sled_agent_address,
+                &log,
+            );
+
+            let config: DatasetsConfig = match config.clone().try_into() {
+                Ok(config) => config,
+                Err(err) => return Some(err)
+            };
+
+            let result =
+                client.datasets_put(&config).await.with_context(
+                    || format!("Failed to put {config:#?} to sled {sled_id}"),
+                );
+            match result {
+                Err(error) => {
+                    warn!(log, "{error:#}");
+                    Some(error)
+                }
+                Ok(result) => {
+                    let (errs, successes): (Vec<_>, Vec<_>) = result
+                        .into_inner()
+                        .status
+                        .into_iter()
+                        .partition(|status| status.err.is_some());
+
+                    if !errs.is_empty() {
+                        warn!(
+                            log,
+                            "Failed to deploy datasets for sled agent";
+                            "successfully configured datasets" => successes.len(),
+                            "failed dataset configurations" => errs.len(),
+                        );
+                        for err in &errs {
+                            warn!(log, "{err:?}");
+                        }
+                        return Some(anyhow!(
+                            "failure deploying datasets: {:?}",
+                            errs
+                        ));
+                    }
 
-/// For each zone in `all_omicron_zones` that has an associated durable dataset,
-/// ensure that a corresponding dataset record exists in `datastore`.
+                    info!(
+                        log,
+                        "Successfully deployed datasets for sled agent";
+                        "successfully configured datasets" => successes.len(),
+                    );
+                    None
+                }
+            }
+        })
+        .collect()
+        .await;
+
+    if errors.is_empty() {
+        Ok(())
+    } else {
+        Err(errors)
+    }
+}
+
+#[allow(dead_code)]
+pub(crate) struct EnsureDatasetsResult {
+    pub(crate) inserted: usize,
+    pub(crate) updated: usize,
+    pub(crate) removed: usize,
+}
+
+/// For all datasets we expect to see in the blueprint, ensure that a corresponding
+/// database record exists in `datastore`.
 ///
-/// Does not modify any existing dataset records. Returns the number of
-/// datasets inserted.
+/// Updates all existing dataset records that don't match the blueprint.
+/// Returns the number of datasets changed.
```

Comment:
I've made this function do a little more work, and return a more informative error status. Check below for additional tests to validate this behavior (tracking insert vs update vs removals).